index.Rmd

```diff
@@ -194,92 +177,90 @@ plot(trainingPoly, add = TRUE)
 
 ```{r reclassify}
-## Inspect the trainingPoly object
-trainingPoly<-trainingPoly[,c(2:4)] #remove an unused column
+# Inspect the trainingPoly object
+trainingPoly <- trainingPoly[, c(2:4)] #remove an unused column
 trainingPoly
 
 # The 'Class' column is a character but should be converted to factor 
 summary(trainingPoly$Class)
 trainingPoly$Class <- as.factor(trainingPoly$Class)
 summary(trainingPoly$Class)
-
 ```
 
 ```{r}
 # We can make a new 'Code' column by converting the factor levels to integer by using the as.numeric() function,
 trainingPoly$Code <- as.numeric(trainingPoly$Class)
+
 # Inspect the new 'Code' column
 summary(trainingPoly$Code)
 ```
 
 ```{r}
-## Plotting
-# Define a colour scale for the classes (as above)
-# corresponding to: cropland, forest, wetland
+# Define a colour scale for the classes (as above) corresponding to: cropland, forest, wetland
 cols <- c("orange", "dark green", "light blue")
-### Superimpose training polygons (colour by class) onto NDVI plot
+
+# Superimpose training polygons (colour by class) onto NDVI plot
 plot(ndvi)
-plot(trainingPoly["Class"], add = TRUE, pal=cols)
-## Add a customised legend
-legend("topright", legend=c("cropland", "forest", "wetland"), fill=cols, bg="white")
+plot(trainingPoly["Class"], add = TRUE, pal = cols)
+
+# Add a customised legend
+legend("topright", legend = c("cropland", "forest", "wetland"), fill = cols, bg = "white")
 ```
 
 
-Our goal in preprocessing this data is to have a table of values representing all layers (covariates) with *known* values/classes. To do this, we will first need to know the values of the covariates at our training polygon locations. We can use `extract` function of raster package for this. Next we convert these data to a `data.frame` representing all training data.
+Our goal in preprocessing this data is to have a table of values representing all layers (covariates) with *known* values/classes. To do this, we will first need to know the values of the covariates at our training polygon locations. We can use `extract` function of terra package for this. Next we convert these data to a `data.frame` representing all training data.
 
 ```{r include=TRUE, results='hide'}
-## Extract pixel values below the polygons
-trainingData<-extract(covs, trainingPoly)
-# Check structure
-str(trainingData)
-# It is a matrix (rows are pixels, columns are bands) per polygon
-# Convert each matrix to a data.frame and add the class column from the polygons
-valuetable <- lapply(1:length(trainingData), function(polygonID) cbind.data.frame(trainingData[[polygonID]], Class=trainingPoly$Class[polygonID]))
-# Unlist into one long data.frame
-valuetable <- do.call(rbind, valuetable)
+# Extract pixel values below the polygons into a dataframe
+trainingData <- extract(covs, trainingPoly)
 
+# Add a column specifying the class based on the polygon ID
+trainingData$Class <- trainingPoly$Class[trainingData$ID]
```

Comment:
Nice! Much more elegant solution :) Students always got confused about this part in the past.